### PR TITLE
htlcswitch: add an always on mode to htlc interceptor

### PR DESCRIPTION
**Pull request by joostjager**
Adds a flag that makes htlc interception a requirement. The switch will queue htlcs as long as there is no interceptor connected. When the stream disconnects, htlcs waiting for resolution will not resume automatically.

Supersedes #6165 

Solves #6161

### CODE REVIEW

File: htlcswitch/interceptable_switch.go
Comment (joostjager) [inline htlcswitch/interceptable_switch.go:45]: To prevent startup order issues, I think it is key to have this boolean here in the interceptable switch.

File: htlcswitch/interceptable_switch.go
Comment (joostjager) [inline htlcswitch/interceptable_switch.go:180]: Upon further inspection, it seems that `linkQuit` is not worth the hassle to thread through.

File: htlcswitch/interceptable_switch.go
Comment (joostjager) [inline htlcswitch/interceptable_switch.go:184]: Synchronizing with the main loop. Should be very light if there's no interceptor.

File: lnrpc/routerrpc/forward_interceptor.go
Comment (joostjager) [inline lnrpc/routerrpc/forward_interceptor.go:59]: `forwardInterceptor` can be cleaned up a lot now that it doesn't need the synchronization anymore to manage the hold list.

File: htlcswitch/interceptable_switch.go
Comment (joostjager) [inline htlcswitch/interceptable_switch.go:48]: This is the main change. `holdForwards` are now kept here. This prevents complexities with `InterceptableSwitch` and `forwardInterceptor` both trying to do things with held htlcs.

File: htlcswitch/interceptable_switch.go
Comment (joostjager) [inline htlcswitch/interceptable_switch.go:71]: Sync with main loop to prevent race conditions. See also https://github.com/lightningnetwork/lnd/pull/6165#discussion_r791431268

**Comment by joostjager** [comment]
@carlaKC @champo I tried a few variations of the always on interceptor and ended up with what's in this PR. Tried to address the outstanding comments in https://github.com/lightningnetwork/lnd/pull/6165

File: htlcswitch/interceptable_switch.go
Comment (joostjager) [inline htlcswitch/interceptable_switch.go:262]: Need to think how the onchain interception flow (https://github.com/lightningnetwork/lnd/pull/6219) can connect to this.

File: htlcswitch/interceptable_switch.go
Comment (joostjager) [reply htlcswitch/interceptable_switch.go:154]: Discussed offline, this isn't ideal because it can create a complicated restart loop if the interceptor itself also restarts when lnd goes down.

File: htlcswitch/interceptable_switch.go
Comment (joostjager) [reply htlcswitch/interceptable_switch.go:186]: Discussed offline that this will be a follow up PR

File: htlcswitch/interceptable_switch.go
Comment (joostjager) [reply htlcswitch/interceptable_switch.go:291]: I rebased #6219 to demonstrate how it can be done.

File: htlcswitch/interceptable_switch.go
Comment (champo) [reply htlcswitch/interceptable_switch.go:285]: This effectively removes the feature entirely, doesn't it? Since packets always come through here. I'm not sure about the merits `linkQuit` has, or lacks, but it IMO this is not the PR to kill it or the way to do it. If desired, it should be removed completely from htlcSwitch rather than silently disused.

**Comment by champo** [comment]
This is a great improvement! Moving the logic to interceptable switch makes a lot of sense.

File: server.go
Comment (joostjager) [reply server.go:1819]: It is the other way around: link -> interceptableSwitch -> htlcSwitch

File: htlcswitch/interceptable_switch.go
Comment (joostjager) [reply htlcswitch/interceptable_switch.go:162]: done

File: htlcswitch/interceptable_switch.go
Comment (joostjager) [reply htlcswitch/interceptable_switch.go:220]: Good point. Converted to log and carry on.

File: htlcswitch/interceptable_switch.go
Comment (joostjager) [reply htlcswitch/interceptable_switch.go:191]: Excellent point. @carlaKC also expressed concerns about this blocking behavior.

Implemented your suggestion, which is indeed simpler and cleaner.

**Comment by alexbosworth** [comment]
could it reject all htlcs when an interceptor is not connected rather than queue them?

**Comment by joostjager** [comment]
>could it reject all htlcs when an interceptor is not connected rather than queue them?

That is a dangerous setup. The interceptor may have signaled to settle an htlc. Then lnd goes down for some reason, replays the packet but finds no interceptor. It will then fail the htlc while the interceptor assumes it has been settled.

**Comment by alexbosworth** [comment]
> > could it reject all htlcs when an interceptor is not connected rather than queue them?
> 
> That is a dangerous setup. The interceptor may have signaled to settle an htlc. Then lnd goes down for some reason, replays the packet but finds no interceptor. It will then fail the htlc while the interceptor assumes it has been settled.

I mean new HTLCs, not ones that were already received, or is that what this does do?

**Comment by joostjager** [comment]
Got it. Yes, I think that could be safe. Not sure though if we have the (persisted) information somewhere that indicates whether the htlc is new or not.

We have the plan to do a height watcher in a follow-up pr that fails htlcs when they get close to the broadcast deadline. This has sort of the same effect when the interceptor doesn't connect anymore, only delayed.

Are concerns about force-closes the reason why you ask?

**Comment by alexbosworth** [comment]
> We have the plan to do a height watcher in a follow-up pr that fails htlcs when they get close to the broadcast deadline. This has sort of the same effect when the interceptor doesn't connect anymore, only delayed.

> Are concerns about force-closes the reason why you ask?

I think force closes are definitely a concern due to closing costs but also the case that excessively holding HTLCs start to represent a cost to the holding party. I think failing HTLCs is a simpler failure mode than holding other people's pending payments open

**Comment by joostjager** [comment]
The always-on mode is intended for interceptors that are high availability. Given that, dangling htlcs should be an edge case.

**Comment by alexbosworth** [comment]
Is the queue bounded by the HTLC slots? The other concern is that if the interceptor failed that HTLCs would queue endlessly?

File: htlcswitch/interceptable_switch.go
Comment (joostjager) [reply htlcswitch/interceptable_switch.go]: Added commit deleting `linkQuit`.

File: htlcswitch/interceptable_switch.go
Comment (joostjager) [inline htlcswitch/interceptable_switch.go:259]: >Is the queue bounded by the HTLC slots? The other concern is that if the interceptor failed that HTLCs would queue endlessly?

Responding in a thread, to not take up so much space in the main issue timeline.

Htlcs only end up in the queue after they've been locked in on the incoming channel. So there is a bound on the incoming side.

**Comment by joostjager** [comment]
Added commit that fails fresh packets if the interceptor is required but not connected. Interested to hear the opinion of reviewers on this addition.

File: htlcswitch/interceptable_switch.go
Comment (joostjager) [reply htlcswitch/interceptable_switch.go]: I also deleted a complicated test case `TestChannelLinkShutdownDuringForward` testing `linkQuit`. To me it seems that the case where the switch is completely blocked isn't realistic, but maybe someone with more context can weigh in.

File: htlcswitch/interceptable_switch.go
Comment (carlaKC) [inline htlcswitch/interceptable_switch.go:180]: This method doesn't really make sense to me anymore? We're using it for both set and unset (interceptor=nil) but we log that we've disconnected? Also on reconnect of a new interceptor, we resume everything that's held?

File: htlcswitch/interceptable_switch.go
Comment (carlaKC) [inline htlcswitch/interceptable_switch.go:146]: Worth adding a comment here that we set/unset, took me a second to grok.

File: htlcswitch/interceptable_switch.go
Comment (carlaKC) [reply htlcswitch/interceptable_switch.go:180]: Ok, cleared up in next commit. I think we could move the nil check + logging so that the progression is clearer.

File: htlcswitch/switch_test.go
Comment (carlaKC) [inline htlcswitch/switch_test.go:3380]: Add a case for clearing out of forwarded packets when we're not running in always-on mode and connect/disconnect/reconnect?

File: htlcswitch/switch_test.go
Comment (carlaKC) [inline htlcswitch/switch_test.go:3409]: Add timeouts to these receives? We don't have a `Guard` fn so these will hang if the test fails.

**Review by carlaKC — COMMENTED** [review]
ACK the approach of not holding new packets when we're disconnected, nice middle ground for the offline interceptor case (and pretty simple to implement).

File: htlcswitch/interceptable_switch.go
Comment (joostjager) [reply htlcswitch/interceptable_switch.go:88]: added

File: htlcswitch/interceptable_switch.go
Comment (joostjager) [reply htlcswitch/interceptable_switch.go:146]: Added

File: htlcswitch/switch_test.go
Comment (joostjager) [reply htlcswitch/switch_test.go:3409]: Added

**Comment by champo** [comment]
I think failing HTLCs when the interceptor is disconnected is not a good idea from a reliability stand point given that only one interceptor can be registered at a given time. It means deploys or restarts of an interceptor are always visible downtime to users or if you use the interceptor to enforce some sort of traffic policy it means unwanted HTLCs might be handled automatically. Always holding HTLCs + height based failure IMO makes it safe to use the API and easier to reason about what's going on from the implementors point of view.

**Comment by alexbosworth** [comment]
A compromise could be: don't hold HTLCs longer than x blocks when interceptor is disconnected?

**Comment by joostjager** [comment]
If the interceptor is down, isn't it preferred to present the user with a failure rather than holding on to the htlc and locking capital along the route for an extended period of time (several blocks)?

**Comment by alexbosworth** [comment]
Yeah I'm saying hold capital for the least possible amount of time

**Comment by champo** [comment]
I guess that depends on whether you expect the interceptor to have high uptime or not. The question can also be: if the intercptor will come right back up, why present the user with a failure? That said, this can be easily customized with a small patch, I'll defer to the mantainers opinions.

**Comment by joostjager** [comment]
What would be a reasonable time for this? To just cover deployment windows, it doesn't need to be long. Definitely not until getting close to htlc expiry.

File: htlcswitch/switch_test.go
Comment (joostjager) [reply htlcswitch/switch_test.go:3421]: Added. This was pre-existing though.

**Comment by joostjager** [comment]
I think a watcher that does the following two things should cover it all:

* Watches block height for all pending intercepts and auto-fails before the channel is force-closed
* Keeps a timer for _new_ pending intercepts and auto-fails if no interceptor connects in time.

File: htlcswitch/interceptable_switch.go
Comment (joostjager) [reply htlcswitch/interceptable_switch.go]: More background: https://github.com/lightningnetwork/lnd/commit/f84cd14b12d78c29a4395088768d8cd6d4cf875d

I realized now that `linkQuit` isn't such a problem. If we are in the resume action originating from the interceptor, we don't need to worry about deadlocks because it's a different goroutine. The `ForwardPackets` call can take a nil `linkQuit` in that case.

File: htlcswitch/interceptable_switch.go
Comment (carlaKC) [inline htlcswitch/interceptable_switch.go:296]: nit: add log here, will probably be helpful for debugging one day.

**Review by carlaKC — APPROVED** [review]
Great refactor and nice solution to the "always on" question.

File: htlcswitch/interceptable_switch.go
Comment (joostjager) [reply htlcswitch/interceptable_switch.go:296]: added

**Review by champo — APPROVED** [review]

File: htlcswitch/interceptable_switch.go
Comment (joostjager) [reply htlcswitch/interceptable_switch.go:404]: This code is no longer running on the link goroutine. The original reason was to prevent a deadlock with switch and link. I think the deadlock can't happen in this case.

**Comment by joostjager** [comment]
Rebased

**Comment by bhandras** [comment]
Taking over review from @yyforyongyu to balance his load a bit.

File: htlcswitch/interceptable_switch.go
Comment (bhandras) [inline htlcswitch/interceptable_switch.go:223]: This is preexisting but looping through `s.holdForwards` may change the order of packets.

File: htlcswitch/interceptable_switch.go
Comment (bhandras) [inline htlcswitch/interceptable_switch.go:205]: Similarly, this may change the order of forwards.

File: htlcswitch/interceptable_switch.go
Comment (joostjager) [reply htlcswitch/interceptable_switch.go:223]: Interesting. Do you think there is a point in guarantees about ordering? I am not sure to what extend order is maintained elsewhere. For example when update_add messages come in in the link. Also a replay from the on-chain flow to the interceptor (next pr) just cuts through whatever order there may have been.

Comment (bhandras) [reply htlcswitch/interceptable_switch.go:223]: Had to dig a bit deeper to fully understand how the forwarding works. There doesn't seem to be any risk in reordering since the interceptor sits right before the Switch.

File: htlcswitch/switch_test.go
Comment (bhandras) [inline htlcswitch/switch_test.go:3452]: To avoid nondeterministic behavior, I'd probably switch the order of `assertNumCircuits` and `assertOutgoingLinkReceive` (even though logically the order is right) since there's a chance that the above call to `ForwardPackets` will be after `assertNumCircuits`. There's a 1 second wait in `assertOutgoingLinkReceive` that makes the ordering correct (with a high chance). Alternatively we could add a forced 1 sec sleep before the asserts.

File: cmd/lncli/cmd_intercept.go
Comment (bhandras) [inline cmd/lncli/cmd_intercept.go:1]: Please don't forget to remove this commit.

**Review by bhandras — APPROVED** [review]
LGTM :tada:

File: cmd/lncli/cmd_intercept.go
Comment (joostjager) [reply cmd/lncli/cmd_intercept.go:1]: Good one! Removed.

**Comment by joostjager** [comment]
Who can hit the merge button on this triple green check pr? There is more work to do within this theme!

**Pull request merged**
